[IMP] deprecate server-side t-raw & remove occurrences

xmo-odoo opened the pull request:
`t-raw` is a relatively regular source of issues, not just because it
gets misused (though that happens) but because it's a form of action
at a distance: a commit adds a `t-raw` with a checked input, this gets
validated and merged, then the *input* gets modified or overridden or
... and the devs & reviewers don't realize it will end up in a
`t-raw`... and we've got an injection vector.

The markupsafe package (on which we already have a hard dependency)
turns out to have a solution for that: if the parameter to
`markupsafe.escape` has a `__html__` method, it will just call that.

It also provides `markupsafe.Markup` which defines exactly one method,
and furthermore will automatically escape the parameters to its
methods (which can be a bit disconcerting for some) as well as
formatting parameters e.g.

    Markup("foo %s %s") % (a, b)

will automatically escape `a` and `b` before they're merged into the
format string, and will return a markup. Obviously that doesn't work
the other way around (because `str.__mod__` and `str.format` always
return a string). It does work for concatenation though.

This means if we can use `markupsafe.escape` as the standard escape
function (which is a good idea since werkzeug [is deprecating
`werkzeug.utils.escape`
anyway](https://github.com/pallets/werkzeug/issues/1758)) we can
leverage this property by requiring the *sources* to be properly
tagged for them to be directly injectible, this means in the scenario
above the payload will either be properly escaped OOTB or it will
break the existing system, but either way it should not be a
*security* risk. And we can easily flag mentions of `Markup` or
`__html__` for security review.

This PR updates `qweb` to use `markupsafe.escape` as its escaping
function, and adds a warning to `t-raw`. It also updates a number of
APIs to make them markup-safe by default:

* html_escape (that's actually what `markupsafe.escape` does on all of its own)
* html_sanitize
* the output of HTML fields
* scriptsafe JSON
* nl2br (which internally escapes its input if necessary)
* qweb markup bodies e.g. `<t t-set="thing">body</t>` the value is markup-safe
* qweb *output*, however `qweb._render` returns *binary data*, for the
  most part this is invisible because we keep using `to_text` and
  `ustr` everywhere and those decode bytes to text automatically, but
  `Markup` inherits from `str` so we can't just have qweb return a
  `Markup`, it generates garbage. Therefore it returns
  `MarkupSafeBytes`

`to_text` was also updated to let `Markup` instances pass through
unmolested. `ustr`, however, was not.

New wrapper types:

`MarkupSafeBytes` is used as the output of `qweb._render`. It extends
`bytes` as `Markup` extends `str`, though it doesn't redefine all the
utility escaping `Markup` does. For the most part:

* It decodes to a `Markup`.
* It defines an `__html__` which decodes to and returns a `Markup`.

`_ScriptSafe` is the new output of the scriptsafe JSON module. It
defines its own `__html__` which applies script-safe escaping at point
of use, it doesn't use the `Markup` wrapper to avoid issues of
unescaping and double-escaping related to attributes, with which
script-safe escaping is not compatible.

A few other notes:

* `to_text` was updated to let `Markup` objects pass through
  unmodified, `ustr` was not updated thus.
* The `Text` fields used through `t-raw` were converted to `Html(...,
  sanitized=False)` (at least when they could not be sanitized
  e.g. header contents and friends).
* Digest was completely untested, but the test I added is not entirely
  reliable, sometimes records go missing and I'm not really sure why.
* Running Odoo with `-b` might be a good idea: while Python 3 removed
  implicit conversions between text and bytes, this means
  `str(b'foo')` now generates the value `"b'foo'"`. Because `Markup`
  extends `str` it has essentially the same behaviour, and passing
  *bytes* to `Markup` is basically a garbage generator.

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/68072).

xmo-odoo commented:
Sarry about the noise I wanted to create it as draft so I could ping @tde-banana-odoo, @odony and @mart-e but I clicked the wrong button.

JKE-be commented on addons/web/views/report_templates.xml:
```diff
@@ -266,50 +266,47 @@
-                                    y[j].textContent = vars[x[i]];
-                            }
+                <script>
+                    function subst() {
```
why not keep t-if='subst' to avoid to have this code when not called ?

xmo-odoo replied:
Sure, why not.

xmo-odoo commented on addons/website/controllers/main.py:
```diff
@@ -137,7 +137,12 @@ def country_infos(self, country, **kw):
     @http.route(['/robots.txt'], type='http', auth="public", website=True, sitemap=False)
     def robots(self, **kwargs):
-        return request.render('website.robots', {'url_root': request.httprequest.url_root}, mimetype='text/plain')
+        return request.make_response(f'''\
```
> Why not keep the view ?

Seemed to be of limited worth given `robots_txt`.

> It will break some use case like custom view enabled (e.g. after a duplicate) to replace default sitemap

Is that really a thing?

> Or a module that disallow Semrush by default on all dbs.

?

> Remove it will make it impossible except with hack to alter the robots_text field... and request.httprequest.url_root no more available in your custo

What do you mean no more available? That you can't remove it anymore?

JKE-be replied:
> What do you mean no more available?

You don't have access to request.httprequest.url_root in robots_txt field.
So if you have a part of your website that is not on Odoo (e.g. you proxypass /xxx to another system) you cannot provide extra sitemap file in your robotx.txt

> Is that really a thing?
I suppose, since I use it myself ;)
On duplicate db for testing purpose, I enable a custom view to remove site map by a disallow all. But I can probably use a nginx config or other thing, but I just don't see the interest to remove the 'feature' that work as expected until now, to have something less customizable and no less risky... except if I miss a security hole here...

xmo-odoo replied:
OK fair enough.

> except if I miss a security hole here...

Nah it was just inconvenient.

xmo-odoo commented:
TODO second commit: under the impetus of @mart-e we tried to find out why markupsafe uses numeric entities for `"` and `'`. The answer is at pallets/jinja@f35e28154f6ee9269beec6864136dbb9f230af7d:

> some documentation improvements, jinja escapes " and ' now, both into charpoints and **no named entities for html 3.2 support** :wink:

So add that to the commit message.

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+from unittest.mock import patch
+import random
+
+from dateutil.relativedelta import relativedelta
```
I am really sad about your import ordering and spacing.

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+from odoo.tests import TransactionCase
+
+
+class TestDigestStuff(TransactionCase):
```
TestDigest should be sufficient, isn't it ?

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+    def test_digest_numbers(self):
+        self._setup_messages()
+
+        d = self.env['digest.digest'].create({'name': "My Digest", 'kpi_mail_message_total': True})
```
I would suggest using meaningful variable name

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+        d = self.env['digest.digest'].create({'name': "My Digest", 'kpi_mail_message_total': True})
+
+        # subscribe the demo user so at least one mail gets sent
+        demo = self.env.ref('base.user_demo')
```
Make a specific user for tests, better than relying on existing data (there are standard users in mail common, a standard employee for example).

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+            d.action_send()
+        self.assertTrue(body, "a mail has been created for the digest")
+
+        kpi_values = html.fromstring(body).xpath('//*[hasclass("kpi_cell")]//*[hasclass("kpi_value")]')
```
Is it a reliable test ? I mean, is it working for example when other KPIs are involved (like having crm and digest) ? Seems a bit low-level and crash-test.

xmo-odoo replied:
It's always going to be low-level since the entire point is to test that the mail is correctly generated (without double-escaping content) and that mail is not exactly conducive to programmatic manipulations. I guess hooking into the kpi_fullname to check that this is the correct "field" could help but only to the extent that the field's label does not change. The alternative is to go and edit the template itself in order to add the field name itself in the output somewhere.

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+            body = self.body_html
+            return _original(self, *args, **kwargs)
+
+        with patch.object(MailMail, 'send', autospec=True, wraps=MailMail, side_effect=send):
```
Why not using existing mail tools and mocks ? I know there are maybe not the cleanest but I would prefer using it than having yet-another-mock-not-completely-like others.

xmo-odoo replied:
Because it didn't work when I tried and I've no idea which one is supposed to intercept message bodies? The one I tried would store the MailMessage objects, but since the mails are auto_delete by the time it returned the mails had been deleted and there was no way to inspect the body.

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+            ['3', '8', '15']
+        )
+
+    def _setup_messages(self):
```
Would move to a setUp .

xmo-odoo replied:
Would not.

tde-banana-odoo replied:
Would r- then.

xmo-odoo replied:
Don't care, it's not a setup for multiple tests, it has no reason to live in a setUp method.

tde-banana-odoo commented on addons/digest/tests/test_digest.py:
```diff
@@ -0,0 +1,75 @@
+        """
+        self.env['mail.message'].search([]).unlink()
+        now = fields.Datetime.now()
+        partners = self.env['res.partner'].search([])
```
Meh. Actually why do you need partners at all ?

xmo-odoo replied:
To post messages on them?

tde-banana-odoo replied:
Arf. Can't you post on some newly creted mail.channel instead ? Looks like "more real use case" oriented.

xmo-odoo replied:
Does that actually work? The message KPI looks for subtype "mt_comment" and message type "comment" or "email".

tde-banana-odoo commented on addons/link_tracker/models/mail_render_mixin.py:
```diff
@@ -3,6 +3,7 @@
 
 import re
 
+import markupsafe
```
HumanLinter: markupsafe should be imported before re

xmo-odoo replied:
Nope, `re` is part of the standard library, markupsafe is not.

tde-banana-odoo replied:
Ah !

xmo-odoo replied:
But the other imports are a mess. PyCharm basically inserted MarkupSafe at the first existing non-stdlib import, but there are more stdlib imports afterwards. I should probably rewrite the entire list of imports here, but I would've expected pylint to  complain.

tde-banana-odoo commented on addons/mail/models/mail_render_mixin.py:
```diff
@@ -226,11 +230,11 @@ def _prepend_preview(self, html, preview):
+            html_preview = Markup("""
                 <div style="display:none;font-size:1px;height:0px;width:0px;opacity:0;">
-                  {tools.html_escape(preview)}
+                   {}
```
Mmmh ? Strange formatting isn't it ? Can't we use standard name-based format ?

xmo-odoo replied:
Do you mean `%`? Switching to str.format was just less change from the original f-string.

tde-banana-odoo commented on addons/mass_mailing/wizard/mail_compose_message.py:
```diff
@@ -78,7 +78,7 @@ def get_mail_values(self, res_ids):
                 }
                 if mail_values.get('body_html') and mass_mail_layout:
-                    mail_values['body_html'] = mass_mail_layout._render({'body': mail_values['body_html']}, engine='ir.qweb', minimal_qcontext=True)
+                    mail_values['body_html'] = mass_mail_layout._render({'body': mail_values['body_html']}, engine='ir.qweb', minimal_qcontext=True).decode()
```
Why this change ? Looking at number of rendering we do in various places, did we miss some encode/decode elsewhere ?

xmo-odoo replied:
See [the commit message](https://github.com/odoo/odoo/pull/68072/commits/d254075871ebf4e9a25d8fd1e3fe9d60d491a2d2).

Most such situations are implicitly decoded because they go through a to_text/ustr/whatever, this one very specifically is not.

tde-banana-odoo replied:
I have to say your message does not really helps dumb people like me understanding the real issue. What is this "Running with `-b`" ?

xmo-odoo replied:
> What is this "Running with -b" ?

`python -b <...>`

> I have to say your message does not really helps dumb people like me understanding the real issue.

```python
>>> rendered = b'<div>whee</div>'
>>> str(rendered)
"b'<div>whee</div>'"
>>> Markup(rendered)
Markup("b'<div>whee</div>'")
```

tde-banana-odoo commented on addons/website_sale/views/product_views.xml:
```diff
@@ -125,6 +125,28 @@
     </record>
 
+    <!-- Product ribbon -->
+    <record id="product_ribbon_form_view" model="ir.ui.view">
```
Side note: a new view ?

xmo-odoo replied:
Yes.

tde-banana-odoo replied:
Side note 2: Why ?

JKE-be replied:
To allow to edit the field 'html' (that was of type Text and now Html) as a text field.
By default, the auto generated view will use widget='html' without this one

:thinking: strange name for a field ... ;)

tde-banana-odoo commented on odoo/tools/mail.py:
```diff
@@ -4,6 +4,8 @@
 import collections
 import logging
+
+import markupsafe
```
HumanLinter is sad about your import oredering and spacing (unnecessary blank like before and missing one after)

xmo-odoo replied:
HumanLinter is wrong.

tde-banana-odoo replied:
Actually human linter is confused as generally odoo imports are not looking like standard lib imports. Sad humain linter.

tde-banana-odoo reviewed (commented):
Blorks ! Looking mainly at Marketing Apps, added some comments.

JKE-be commented on addons/website_sale/views/templates.xml:
```diff
@@ -1849,7 +1849,7 @@
                         <div t-attf-class="carousel-item h-100#{' active' if product_image_first else ''}">
                             <div t-if="product_image._name == 'product.image' and product_image.embed_code" class="d-flex align-items-center justify-content-center h-100 embed-responsive embed-responsive-16by9">
-                                <t t-raw="product_image.embed_code"/>
+                                <t t-esc="product_image.embed_code"/>
```
This fields probably need to be declared as html ;) 
![image](https://user-images.githubusercontent.com/4848797/114015954-9f35f180-986a-11eb-9c44-56195b288e8f.png)

xmo-odoo replied:
ndeed, I probably got confused between the embed_code from website_slides and the one from website_sales and missed one of the two.

JKE-be commented on addons/website_sale/views/templates.xml:
```diff
@@ -213,7 +213,7 @@
             <t t-set="text_color" t-value="td_product['ribbon']['text_color']"/>
             <t t-set="bg_class" t-value="td_product['ribbon']['html_class']"/>
-            <span t-attf-class="o_ribbon #{bg_class}" t-attf-style="#{text_color and ('color: %s; ' % text_color)}#{bg_color and 'background-color:' + bg_color}" t-raw="td_product['ribbon']['html'] or ''"/>
+            <span t-attf-class="o_ribbon #{bg_class}" t-attf-style="color: {{text_color or 'inherit'}}; background-color: {{bg_color or 'inherit'}}" t-esc="td_product['ribbon']['html'] or ''"/>
```
it is not the same ! 
https://drive.google.com/file/d/1pXgTvjDjIEj2diYRAvjN31pMHJMZRn9n/view

Elkasitu commented on addons/website_sale/__manifest__.py:
```diff
@@ -7,7 +7,7 @@
     'summary': 'Sell your products online',
     'website': 'https://www.odoo.com/page/e-commerce',
-    'version': '1.1',
+    'version': '1.2',
```
is this because of the new view?

xmo-odoo replied:
Should not be necessary for that, I think I removed a view at one point but then might have reverted it.

Elkasitu commented on odoo/fields.py:
```diff
@@ -1683,6 +1683,7 @@ class Html(_String):
     """
     type = 'html'
     column_type = ('text', 'text')
+    column_cast_from = ('varchar',)
```
i don't know how you plan to rebase your commits, but I think it'd make more sense to have this change in the feature commit

xmo-odoo replied:
True, I probably have this one here because it was the first commit where I needed to replace a `Char` field with an `Html` one.

Elkasitu commented on addons/portal/views/portal_templates.xml:
```diff
@@ -302,9 +302,9 @@
-                                    <a t-att-href="'#' + searchbar_inputs[input]['input']"
-                                        t-attf-class="dropdown-item#{search_in == searchbar_inputs[input]['input'] and ' active' or ''}">
-                                        <span t-raw="searchbar_inputs[input]['label']"/>
+                                    <a t-att-href="'#' + input_value['input']"
```
where does `input_value` come from?

xmo-odoo replied:
`t-foreach` creates a bunch of in-scope names based on the `t-as`.

When iterating a mapping, the key goes into `{as}` and the value goes into `{as}_value` (for non-mapping, both get the same).

Elkasitu reviewed (approved):
some small remarks, otherwise LGTM

xmo-odoo commented on odoo/addons/base/models/qweb.py:
```diff
@@ -745,18 +784,27 @@ def _append_attributes(self):
                         ]
                     ),
                     body=[
+                        # append(' '')
```
Likely my editor auto-doubling the quote and then I didn't notice.

xmo-odoo commented:
As apparently discussed during the gooroo meeting, added `t-out` as alias to `t-esc` (the other way around really, `t-esc` now calls `t-out` internally), and updated the PR's commits to replace the t-raws by t-out instead of t-esc.

xmo-odoo commented on odoo/addons/base/models/qweb.py:
```diff
@@ -1702,26 +1763,34 @@ def _compile_expr0(self, expr):
         """ Parses the provided format string and compiles it to a single
         expression ast, uses string concatenation via "+".
+
+        Literal sections of the format are wrapped in Markup.
```
It's not anymore.

xmo-odoo commented on odoo/addons/base/models/qweb.py:
```diff
@@ -691,15 +709,36 @@ def _values_var(self, varname, ctx):
+    # which can not appear as-is in an attribute and must be escaped.
+    #
+    # The only possible option I can see here is to unescape `Markup` values
+    # before re-escaping them.
```
Remove comment as better fix was implemented.

xmo-odoo commented on odoo/addons/base/models/qweb.py:
```diff
@@ -1673,17 +1731,20 @@ def _get_widget(self, value, expression, tagName, field_options, options, values
+                    func=ast.Name(id='to_text', ctx=ast.Load()),
+                    args=[self._compile_expr(expr)], keywords=[],
+                )
+            ], keywords=[]
```
Note: JS behaves slightly differently here, non-text values can pass through as-is and are only stringified (with null/undefined being replaced by empty strings) in the `_append_attribute` equivalent. This is for the text-based compiler, I've not looked at the other one.

xmo-odoo reviewed (commented):
Reminder of a few bits I've forgotten.

xmo-odoo commented:
@robodoo rebase-merge

xmo-odoo commented:
@robodoo override=ci/security

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

xmo-odoo commented:
@robodoo r+
